commons/src/test/java/io/aiven/kafka/tieredstorage/commons/UniversalRemoteStorageManagerTest.java

```diff
@@ -0,0 +1,348 @@
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+
+import org.apache.kafka.common.TopicIdPartition;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.server.log.remote.storage.LogSegmentData;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentId;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+import org.apache.kafka.server.log.remote.storage.RemoteStorageException;
+import org.apache.kafka.server.log.remote.storage.RemoteStorageManager;
+
+import io.aiven.kafka.tieredstorage.commons.manifest.index.ChunkIndex;
+import io.aiven.kafka.tieredstorage.commons.security.EncryptionKeyProvider;
+
+import com.github.luben.zstd.Zstd;
+import org.apache.commons.io.input.BoundedInputStream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.JsonNode;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class UniversalRemoteStorageManagerTest extends RsaKeyAwareTest {
+    RemoteStorageManager rsm;
+
+    EncryptionKeyProvider encryptionKeyProvider;
+
+    @TempDir
+    Path tmpDir;
+    Path sourceDir;
+    Path targetDir;
+
+    Path logFilePath;
+    Path offsetIndexFilePath;
+    Path timeIndexFilePath;
+    Path producerSnapshotFilePath;
+    Path txnIndexFilePath;
+
+    static final int SEGMENT_SIZE = 10 * 1024 * 1024;
+    static final Uuid TOPIC_ID = Uuid.METADATA_TOPIC_ID;  // string representation: AAAAAAAAAAAAAAAAAAAAAQ
+    static final Uuid SEGMENT_ID = Uuid.ZERO_UUID;  // string representation: AAAAAAAAAAAAAAAAAAAAAA
+    static final TopicIdPartition TOPIC_ID_PARTITION = new TopicIdPartition(TOPIC_ID, new TopicPartition("topic", 7));
+    static final RemoteLogSegmentId REMOTE_SEGMENT_ID = new RemoteLogSegmentId(TOPIC_ID_PARTITION, SEGMENT_ID);
+    static final long START_OFFSET = 23L;
+    static final RemoteLogSegmentMetadata REMOTE_LOG_METADATA = new RemoteLogSegmentMetadata(
+        REMOTE_SEGMENT_ID, START_OFFSET, 2000L,
+        0, 0, 0, SEGMENT_SIZE, Map.of(0, 0L));
+
+    static final ByteBuffer LEADER_EPOCH_INDEX = ByteBuffer.wrap("leader epoch index".getBytes());
+
+    static final String TARGET_LOG_FILE =
+        "topic-AAAAAAAAAAAAAAAAAAAAAQ/7/00000000000000000023-AAAAAAAAAAAAAAAAAAAAAA.log";
+    static final String TARGET_MANIFEST_FILE =
+        "topic-AAAAAAAAAAAAAAAAAAAAAQ/7/00000000000000000023-AAAAAAAAAAAAAAAAAAAAAA.rsm-manifest";
```

Comment:
Would it make sense to use ObjectKey #164 to build paths here?

Reply:
Most of the time I try to avoid this approach if it's not too cumbersome. It's more convenient, but instead of "this test tests a particular expected behavior of the code" it turns into "this test tests that the code is aligned with another piece of code which may be not correct in the first place"

Reply:
Got it, fair enough